### PR TITLE
manual SDL2 library search

### PR DESCRIPTION
**Pull request by jpcima**
This detects SDL2 without using the FindModule shipped with the library,
which is unreliable. It's the manual search which works in all configurations.

("old way" copied back from libOPNMIDI)
see https://github.com/Wohlstand/libADLMIDI/issues/196#issuecomment-457079484

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -233,6 +233,29 @@ else()
     add_library(ADLMIDI ALIAS ADLMIDI_static)
 endif()
 
+if(WITH_MIDIPLAY OR WITH_ADLMIDI2)
+    find_library(SDL2_LIBRARY SDL2 REQUIRED)
+    message("Found ${SDL2_LIBRARY}")
```
Comment (Flamefire) [inline CMakeLists.txt]: I'd rather see this as whole code as a find module. You could use ours: https://github.com/Return-To-The-Roots/s25client/blob/master/cmake/Modules/FindSDL2.cmake

File: CMakeLists.txt [hunk 2]
```diff
@@ -233,6 +233,29 @@ else()
     add_library(ADLMIDI ALIAS ADLMIDI_static)
 endif()
 
+if(WITH_MIDIPLAY OR WITH_ADLMIDI2)
+    find_library(SDL2_LIBRARY SDL2 REQUIRED)
```
Comment (Flamefire) [inline CMakeLists.txt]: REQUIRED does not exist here

File: CMakeLists.txt [hunk 3]
```diff
@@ -233,6 +233,29 @@ else()
...
+    message("Found ${SDL2_LIBRARY}")
+
+    set(SDL2_NEEDED_LIBRARIES)
```
Comment (Flamefire) [inline CMakeLists.txt]: Why not use standard names? (no `NEEDED` infix)

File: CMakeLists.txt [hunk 4]
```diff
@@ -233,6 +233,29 @@ else()
...
+    if(WIN32)
+        if(MSVC)
+            list(APPEND SDL2_NEEDED_LIBRARIES "${SDL2_LIBRARY}")
+        else()
+            list(APPEND SDL2_NEEDED_LIBRARIES "${SDL2_LIBRARY}" pthread)
```
Comment (Flamefire) [inline CMakeLists.txt]: Rather use something like `find_package(Threads)` and the results of it

File: CMakeLists.txt [hunk 5]
```diff
@@ -233,6 +233,29 @@ else()
...
+        endif()
+    else()
+        list(APPEND SDL2_NEEDED_LIBRARIES "${SDL2_LIBRARY}" pthread)
+        if(NOT HAIKU AND NOT OPENBSD_LOCALBASE)
+            list(APPEND SDL2_NEEDED_LIBRARIES dl)
+        endif()
+        list(APPEND SDL2_NEEDED_LIBRARIES m stdc++)
```
Comment (Flamefire) [inline CMakeLists.txt]: please don't link to stdc++ explicitely

File: CMakeLists.txt [hunk 6]
```diff
@@ -233,6 +233,29 @@ else()
...
+    endif()
+    if(OPENBSD_LOCALBASE)
+        list(REMOVE_ITEM SDL2_NEEDED_LIBRARIES dl)
```
Comment (Flamefire) [inline CMakeLists.txt]: Not required as never added

**Comment by jpcima** [comment]
I rewrote the whole PR @Flamefire thanks to your advice, do you think it's fine?

**Comment by Flamefire** [comment]
Problem now is that you duplicate the conditions used in the subfolders. They might easily become out-of-sync. I'd suggest either:
- Duplicate the target check at the point of usage. This is kinda fine as one sees why things are done
- Create a very simple find module containing basically your code for ADLMIDI_SDL2 and `find_package` that.
- Create a wrapper function like `function(find_sdl2_targets) find_package(SDL2 ${ARGV}) <your code>` and include+call that: `find_sdl2_targets(REQUIRED) target_link_libraries(... SDL2::SDL2)`

**Comment by jpcima** [comment]
Ok to merge this @Wohlstand ?
It's an annoyance to work from the master branch, because of the cmake failure.
(the case when SDL2 was built with cmake, if I understand right)

**Comment by Wohlstand** [comment]
Looks much better! Yeah, I think, the same set of works are must go into libOPNMIDI's build too

**Pull request merged**